### ISSUE TITLE
GUROBI_RUN.py not working when model stops due to time-out

### ISSUE THREAD

**Issue by MMeketon** [Pyomo/pyomo#365]
The newest GUROBI_RUN.py on Github (https://github.com/Pyomo/pyomo/blob/master/pyomo/solvers/plugins/solvers/GUROBI_RUN.py) does not return a solution when a solution exists but the model stopped due to a time-out.

It does seem to solve the issue of doing a bulk-load of the variables, unlike Version 5.3.  

Please fix this bug.  Right now I'm using a hack to get around this.

**Comment by ghackebeil** [comment]
@MMeketon: Hi Marc. I've run some tests locally on master and this functionality seems to be working. Do you have an example you can share that would help me reproduce the issue you are seeing?

**Comment by ghackebeil** [comment]
Responding based on some additional info that @MMeketon sent me directly:

I think what I see in your code snippet is that you are calling something like `opt.solve(model, ...)` **without** using `load_solutions=False`. This has the effect of the solution being taken out of the results object and automatically loaded into the model, which makes your later call to `model.solutions.load_from(results)` redundant.

This also causes the `aborted` status to trigger the error you are seeing when `load_from` is called rather than just a warning (because it checks if any solutions are available in the results object).

I think what you want to do is

```python
results = opt.solve(model, load_solutions=False, ...)
# do some checking of the statuses on the results object
...
model.solutions.load_from(results)
```

Optionally, if you are annoyed by the WARNING that we print when you load a solution tagged with an `aborted` status, you can do something like:

```python
from pyomo.opt import SolverStatus
...
results = opt.solve(model, load_solutions=False, ...)
if len(results.solution) == 0:
    # uh oh
    ...
else:
    if (results.solver.status == SolverStatus.aborted):
        # addition checks as appropriate
        ...
        if <satisfied>:
            # disable the warning
            results.solver.status = SolverStatus.ok
    else:
        # check for other situations
        ...
    model.solutions.load_from(results)

**Comment by ghackebeil** [comment]
I also want to emphasize that I think it is a bug that calling `model.solutions.load_from(results)` doesn't really alert you when `len(results.solution) == 0`. I'd like to hear what the other developers think about that.

**Comment by jwatsonnm** [comment]
Hard to argue: That's a bug, and an exception should be raised.

**Comment by jsiirola** [comment]
Agreed.  This should at least issue an error, and probably should raise an exception.

**Comment by MMeketon** [comment]
I tested your solution using `load_solutions=False` on the Gurobi solver and it appears to be working.

However, I disagree with the answer of using the `load_solutions=False` option.  In the snippet I sent to Gabe, I only used Gurobi as the solver.  I alternate using CBC, GLPK, XPRESS and Gurobi.  CBC, GLPK and XPRESS all work quite nicely without using the `load_solutions=False` option when the solvers terminate at the max-time-limit.  But not Gurobi.

We should not have to need a different set of options for Gurobi.  

My code looks something like:
```
            if self.solver=="GLPK":
                opt = po.SolverFactory("glpk")
                opt.options["cuts"] = ''
                opt.options["tmlim"]=self.max_seconds_optimizer
                opt.options["mipgap"] = self.optimality_gap
                opt.options["fpump"] = ''
                opt.options["proxy"]  = 120
            elif self.solver.lower()=="xpress":                
                opt = po.SolverFactory("amplxpress")
                opt.options["GOMCUTS"]=0
                opt.options["TREECOVERCUTS"]=0
                opt.options["MAXTIME"]=self.max_seconds_optimizer
                opt.options["MIPRELSTOP"]=self.optimality_gap
                opt.options["heurrootcutfreq"]=2
                opt.options["heurroot"] = 3
                opt.options["LOGFILE"]="CON:"
            elif self.solver=="CBC":
                opt = po.SolverFactory("cbc")
                opt.options["sec"] = self.max_seconds_optimizer
                opt.options["ratio"] = self.optimality_gap
            elif self.solver=="GUROBI":
                opt = po.SolverFactory("gurobi")
                opt.options["TimeLimit"] = self.max_seconds_optimizer
                opt.options["MIPGap"] = self.optimality_gap
                opt.options["NormAdjust"]= 0
                opt.options["DegenMoves"] = 1
                opt.options["Heuristics"] = 0.5
                opt.options["PrePasses"] = 1

            results = opt.solve(self.model, report_timing=True, tee=True)
            print("Solver status: {}".format(results.solver.status))
            print("Termination condition: {}".format(results.solver.termination_condition))

            self.model.solutions.load_from(results)

```

**Comment by ghackebeil** [comment]
First, let me be more clear that you should **not** call `self.model.solutions.load_from(results)` if you are not setting `load_solutions=False`. When you are calling it in your code, the results.solution object is empty (because it was loaded during the opt.solve call because load_solutions defaults to True), so your call is doing nothing except turning a warning into an error when Gurobi exits due to a time limit.

My previous comment was pointing out that it is our fault that it was not clear that your call to load_from was doing nothing (via an exception of some sort).

If you delete the `self.model.solutions.load_from(results)` line (and leave the opt.solve call as it is), then you should not get an error when Gurobi stops with a time limit (you will only get a warning), and the solution will have been automatically loaded.

Using `load_solutions=False` combined with `model.solutions.load_from` should be used only when you want to check statuses prior to loading the solution to avoid an error or warning.

Does this make sense?

Any other differences in behavior are likely due to inconsistencies between the solver interfaces in how different exit conditions are tagged in the results object. For instance, the gurobi solver interface might tag an early termination that includes a feasible solution with SolutionStatus.aborted, whereas the CBC interface might use SolutionStatus.warning. This is definitely something that we need to work on, but this kind of work isn't really prioritized by developers until we have to deal with it for a particular solver on a particular project.

**Comment by MMeketon** [comment]
Thank you for the clarification.  Very helpful.

Please note that there are 11 examples of `solutions.load_from(results)` in the documentation [here.](https://software.sandia.gov/downloads/pub/pyomo/PyomoOnlineDocs.html)  Ten of them do not use the `load_solutions=False` in the `opt.solve` line just above the `load_from` line.  Only in the last example is the option `load_solutions=False` present.  Hence my confusion - it appeared that I always needed the `solutions.load_from(results)`.

**Comment by ghackebeil** [comment]
Thanks for pointing out that documentation! We will get it fixed ASAP.

**Comment by ghackebeil** [comment]
I've opened up new tickets for everything we discussed. I don't think any related fixes are going to make it in to the Pyomo 5.4 release (potentially going out tonight), but I've tagged them as Pyomo 5.5 issues.
